core/domain/exp_jobs_one_off.py

```diff
@@ -728,3 +728,179 @@ def map(item):
+                item.id)
+            return
+
+        versioned_exploration_states = {
+            'states_schema_version': current_state_schema_version,
+            'states': item.content['states']
+        }
+        while current_state_schema_version < target_state_schema_version:
+            try:
+                assert (
+                    versioned_exploration_states['states_schema_version'] ==
+                    current_state_schema_version)
+                exp_domain.Exploration.update_states_from_model(
+                    versioned_exploration_states,
+                    current_state_schema_version,
+                    exp_id)
+                current_state_schema_version += 1
+            except Exception as e:
+                error_message = (
+                    'Exploration snapshot %s failed migration to states '
+                    'v%s: %s' % (
+                        item.id, current_state_schema_version + 1, e))
+                logging.exception(error_message)
+                yield ('MIGRATION_ERROR', error_message.encode('utf-8'))
+                break
+
+            if target_state_schema_version == current_state_schema_version:
+                yield ('SUCCESS', 1)
+
+    @staticmethod
+    def reduce(key, values):
+        if key.startswith('SUCCESS'):
+            yield (key, len(values))
+        else:
+            yield (key, values)
+
+
+class ExpSnapshotsMigrationJob(jobs.BaseMapReduceOneOffJobManager):
+    """A reusable one-time job that may be used to migrate exploration schema
+    versions. This job will load all snapshots of all existing explorations
+    from the datastore and immediately store them back into the datastore.
+    The loading process of an exploration in exp_services automatically
+    performs schema updating. This job persists that conversion work, keeping
+    explorations up-to-date and improving the load time of new explorations.
+
+    NOTE TO DEVELOPERS: Make sure to run ExpSnapshotsMigrationAuditJob before
+    running this job.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationSnapshotContentModel]
+
+    @classmethod
+    def enqueue(cls, job_id, additional_job_params=None):
+        super(ExpSnapshotsMigrationJob, cls).enqueue(
+            job_id, shard_count=64)
+
+    @staticmethod
+    def map(item):
+        exp_id = item.get_unversioned_instance_id()
+
+        latest_exploration = exp_fetchers.get_exploration_by_id(
+            exp_id, strict=False)
+        if latest_exploration is None:
+            yield ('INFO - Exploration does not exist', item.id)
+            return
+
+        exploration_model = exp_models.ExplorationModel.get(exp_id)
+        if (exploration_model.states_schema_version !=
+                feconf.CURRENT_STATE_SCHEMA_VERSION):
+            yield (
+                'FAILURE - Exploration is not at latest schema version', exp_id)
+            return
```

Comment:
Why should an error be raised? Can't we just save the exploration back directly here, since `get_exploration_by_id` already migrated to latest version? Or are you doing that manually when the job is run?

Reply:
Hmm. So my plan was to verify that all the exps are at their latest schema before we run the job (I added that as a pre-check in the job instructions doc). If not then we can run the ExpMigrationJobManager to force a migration before we start the job, so this clause here is just a sanity check.

We can do it the way you suggest, too. But I think I lean more towards this job focusing on migrating (and mutating) just the snapshots only. What do you think?

Reply:
Isn't this check actually needed since otherwise we can end up in a state where snapshots are newer than the latest change? That seems like a potential invariant violation (with regards to data integrity for explorations).

Reply:
The snapshots wouldn't be newer, just their schema version. (Note that all the schema versions would be functionally equivalent anyway, since older ones are migrated on the fly before being used at all.)

Reply:
The way you suggested sounds good with the pre-check and migration, in which case this is just a sanity check that we don't expect to fail. 
I made the comment since I thought this was an expected output after which the migration was to be done, though if the migration is done beforehand and all explorations are validated to be on the latest schema beforehand, it's fine.